# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsContentUtils.cpp
# Commit: 1ae213bfa43e
# Full Hash: 1ae213bfa43e3270689641632f1a89e4c4affb8a
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2022-08-23 02:46:57
# Regressor Bug: 1781129
# File Overlap Count: 1
# Description:
#   Bug 1781129 - Part 3: Remove Shmem overload of GetSurfaceData, r=edgar
#   
#   After the previous changes there was only one consumer left of the Shmem
#   version of GetSurfaceData, which could easily be changed to use BigBuffer,
#   removing the need for that overload.
# ==============================================================================

diff -r dcebc98ea1f1 -r 1ae213bfa43e dom/base/nsContentUtils.cpp
--- a/dom/base/nsContentUtils.cpp	Mon Aug 22 15:38:15 2022 +0000
+++ b/dom/base/nsContentUtils.cpp	Mon Aug 22 15:38:15 2022 +0000
@@ -205,7 +205,6 @@
 #include "mozilla/gfx/Types.h"
 #include "mozilla/ipc/ProtocolUtils.h"
 #include "mozilla/ipc/SharedMemory.h"
-#include "mozilla/ipc/Shmem.h"
 #include "mozilla/net/UrlClassifierCommon.h"
 #include "mozilla/widget/IMEData.h"
 #include "nsAboutProtocolUtils.h"
@@ -8171,62 +8170,12 @@
   }
 }
 
-namespace {
-// The default type used for calling GetSurfaceData(). Gets surface data as
-// raw bigbuffer.
-struct GetSurfaceDataBigBuffer {
-  using ReturnType = Maybe<BigBuffer>;
-  using BufferType = char*;
-
-  ReturnType Allocate(size_t aSize) { return Some(BigBuffer(aSize)); }
-
-  static BufferType GetBuffer(ReturnType& aReturnValue) {
-    return aReturnValue ? reinterpret_cast<char*>(aReturnValue->Data())
-                        : nullptr;
-  }
-
-  static ReturnType NullValue() { return ReturnType(); }
-};
-
-// The type used for calling GetSurfaceData() that allocates and writes to
-// a shared memory buffer.
-struct GetSurfaceDataShmem {
-  using ReturnType = Maybe<Shmem>;
-  using BufferType = char*;
-
-  explicit GetSurfaceDataShmem(IShmemAllocator* aAllocator)
-      : mAllocator(aAllocator) {}
-
-  ReturnType Allocate(size_t aSize) {
-    Shmem shmem;
-    if (!mAllocator->AllocShmem(aSize, &shmem)) {
-      return Nothing();
-    }
-
-    return Some(shmem);
-  }
-
-  static BufferType GetBuffer(const ReturnType& aReturnValue) {
-    return aReturnValue.isSome() ? aReturnValue.ref().get<char>() : nullptr;
-  }
-
-  static ReturnType NullValue() { return ReturnType(); }
-
- private:
-  IShmemAllocator* mAllocator;
-};
-
-/*
- * Get the pixel data from the given source surface and return it as a buffer.
- * The length and stride will be assigned from the surface.
- */
-template <typename GetSurfaceDataContext = GetSurfaceDataBigBuffer>
-typename GetSurfaceDataContext::ReturnType GetSurfaceDataImpl(
-    DataSourceSurface& aSurface, size_t* aLength, int32_t* aStride,
-    GetSurfaceDataContext aContext = GetSurfaceDataContext()) {
+Maybe<BigBuffer> nsContentUtils::GetSurfaceData(DataSourceSurface& aSurface,
+                                                size_t* aLength,
+                                                int32_t* aStride) {
   mozilla::gfx::DataSourceSurface::MappedSurface map;
   if (!aSurface.Map(mozilla::gfx::DataSourceSurface::MapType::READ, &map)) {
-    return GetSurfaceDataContext::NullValue();
+    return Nothing();
   }
 
   size_t bufLen = 0;
@@ -8236,38 +8185,18 @@
       &bufLen);
   if (NS_FAILED(rv)) {
     aSurface.Unmap();
-    return GetSurfaceDataContext::NullValue();
-  }
-
-  // nsDependentCString wants null-terminated string.
-  typename GetSurfaceDataContext::ReturnType surfaceData =
-      aContext.Allocate(maxBufLen + 1);
-  if (GetSurfaceDataContext::GetBuffer(surfaceData)) {
-    memcpy(GetSurfaceDataContext::GetBuffer(surfaceData),
-           reinterpret_cast<char*>(map.mData), bufLen);
-    memset(GetSurfaceDataContext::GetBuffer(surfaceData) + bufLen, 0,
-           maxBufLen - bufLen + 1);
-  }
+    return Nothing();
+  }
+
+  BigBuffer surfaceData(maxBufLen);
+  memcpy(surfaceData.Data(), map.mData, bufLen);
+  memset(surfaceData.Data() + bufLen, 0, maxBufLen - bufLen);
 
   *aLength = maxBufLen;
   *aStride = map.mStride;
 
   aSurface.Unmap();
-  return surfaceData;
-}
-}  // Anonymous namespace.
-
-Maybe<BigBuffer> nsContentUtils::GetSurfaceData(DataSourceSurface& aSurface,
-                                                size_t* aLength,
-                                                int32_t* aStride) {
-  return GetSurfaceDataImpl(aSurface, aLength, aStride);
-}
-
-Maybe<Shmem> nsContentUtils::GetSurfaceData(DataSourceSurface& aSurface,
-                                            size_t* aLength, int32_t* aStride,
-                                            IShmemAllocator* aAllocator) {
-  return GetSurfaceDataImpl(aSurface, aLength, aStride,
-                            GetSurfaceDataShmem(aAllocator));
+  return Some(std::move(surfaceData));
 }
 
 Maybe<IPCImage> nsContentUtils::SurfaceToIPCImage(DataSourceSurface& aSurface) {